### ISSUE TITLE
hledger rewrite ignores negative match

### ISSUE THREAD

**Issue by msmart** [simonmichael/hledger#745]
I'm trying to use `hledger rewrite` to rewrite postings which match certain tags combinations. I would like to run `hledger rewrite` several times without adding additional rewrites. 

To this end, I thought I can add the tag "rewritten" to the rewritten posting and ignore them in the query of the rewrite rule. But in my environment this is not working as I would expect it to.

Given the following sample.ledger file:
```
2018/4/1 * test
    ; details:beer
    a  2
    b 

2018/4/1 * another test
    ; details:food
    a  20
    b 
```
and the following sample.rewrite file 

```
= a tag:details=beer not:tag:rewritten
    a  *-1 ; rewritten:
    c  *1 
```
The `hledger rewrite -f sample.journal -f sample.rewrite| hledger print -f -` gives:

```
2018/04/01 * test     
    ; details:beer    
    a               2 
    b                 
    a              -2    ; rewritten:        
    c               2 

2018/04/01 * another test                    
    ; details:food    
    a              20 
    b                 

```
However, `hledger rewrite -f sample.journal -f sample.rewrite|hledger print -f -|hledger rewrite -f - -f sample.rewrite` gives

```
2018/04/01 * test
    ; details:beer
    a               2
    b
    a              -2    ; rewritten:
    c               2
    a              -2    ; rewritten:
    c               2

2018/04/01 * another test
    ; details:food
    a              20
    b
```
I would have expected that the posting with details:beer and tag:rewritten would not match a second time.

Am I missing something here?

PS: Thanks to all involved in the development of hledger from a happy user.

**Comment by simonmichael** [comment]
Cf http://hledger.org/manual.html#rewrite and the more informative `hledger rewrite -h`. It seems to be simply another way to invoke automated postings (instead of --auto), with the ability to provide rules on the command line (--add-posting) and to show a diff (--diff).

**Comment by msmart** [comment]
thanks for the quick feedback. 

I've tried using automated postings with the --auto option and the files mentioned above on hledger version 1.9 with the following command:
```
hledger -f sample.journal -f sample.rewrite print --auto|hledger -f - -f sample.rewrite print --auto
```

But this gives the same result as the rewrite subcommand:
```
2018/04/01 * test
    ; details:beer
    a               2
    b
    a              -2    ; rewritten:
    c               2
    a              -2    ; rewritten:
    c               2

2018/04/01 * another test
    ; details:food
    a              20
    b
```
On a side note: When removing the account name from the sample.rewrite file like this:
```
= not:c tag:details=beer not:tag:rewritten
    a  *-1 ; rewritten:
    c  *1 
```
```
hledger -f sample.journal -f sample.rewrite print --auto
```
returns
```
2018/04/01 * test
    ; details:beer
    a               2
    b
    a              -2    ; rewritten:
    c               2
    a               2    ; rewritten:
    c              -2

2018/04/01 * another test
    ; details:food
    a              20
    b
```
Which I also did not expect. This possibly might be a good place to start debugging.

For the time being, I'll change my workflow to side step the issue.

Thanks again.

**Comment by simonmichael** [comment]
This is quite hard to follow and requires careful reading. More notes:

- Re ony's [comment](https://github.com/simonmichael/hledger/issues/745#issuecomment-384875666) suggesting adding a transaction comment to the rule. As of hledger-1.10, such a comment is misparsed as another posting (the semicolon has no effect) and you may get an error about a missing amount.

- Another difference between `rewrite` and `print --auto`: rewrite combines all files into one journal before running rules, so rules in one file affect all files. print follows the usual [directive scope rules](http://hledger.org/manual.html#directives), so rules in one file do not affect the other files. As of hledger 1.10 at least.

- It's not great that rewrite and print --auto do almost the same thing, and in slightly different ways. I have intended for rewrite to grow and be able to do more complex things (like rewriting and removing postings). An simpler fix would be to just drop it.

- I'm not able to reproduce either result in msmart's [last comment](https://github.com/simonmichael/hledger/issues/745#issuecomment-384884763), with hledger 1.9 or 1.10.

**Comment by simonmichael** [comment]
- `rewrite QUERY` can apply rules selectively only to the (matched postings in) transactions matched by QUERY (but still prints all transactions). `print --auto QUERY` applies rules to the postings in all transactions, then prints only the transactions matched by QUERY.

**Comment by simonmichael** [comment]
I think this can be closed.
